New service model

slluis opened the pull request:
**REVIEW WARNING: This PR is not 100% complete and will not build since it is part of PR #7158, which changes the document/view architecture. I just created this PR to make the service model changes easier to review.**

**Also, it is easier to review this PR commit by commit.**

This PR implements a new service model for MonoDevelop. Until now we have a mix of approaches for implementing services. Some are implemented as instance classes, some as static classes, and there is no standard way of initializing them.

The new service model has three main goals:

* Homogenize how services are implemented, instantiated, initialized and accessed.
* Support asynchronous and on-demand initialization of services, to reduce and parallelize the of work that needs to be done at IDE startup.
* Improve support for unit testing by allowing discrete service initialization and using mock test services.

The model is very simple:

* The new ServiceProvider class provides methods for registering and getting references to services. 
* Services implement the interface `IService`, either directly or through the abstract class `Service`. The interface defines the async method `Task Initialize(ServiceProvider)` which is invoked the first time a service is requested.

In this new model all services are instances (no more static classes, with the exception of some special cases), and it is possible to get any service using the main service provider available in `Runtime.ServiceProvider`. The `Runtime` class also provides some shortcut methods for getting services. For example:

```
var fontService = await Runtime.GetService<FontService> ();
...
```

The ´GetService()` method can be used to get a service. Services are created on demand, so if the service has not been requested before, it will be created and initialized at this time.

Getting a service is an asynchronous operation (since it may involve initialization). This may not always be convenient, since services also need to be used from methods that are not asynchronous. To make use of services easier MonoDevelop.Ide defines the class IdeServices, which has references to all services that are initialized at startup. Referencing services from IdeServices is safe once the Ide is initialized. In some scenarios (such as in unit tests or when implementing a command line command), IdeServices still can be used, but will only have references to those services that have been initialized.

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/487937.

slluis committed:
Add basic service infrastructure

slluis committed:
Convert several IDE services into Service instances

The new IdeServices class has references to the core IDE services.
Revamped the IDE initialization code. The UI loop is started earlier and
that allows starting the services asynchronously.

Code that previously was referencing static service classes or services
available in IdeApp now reference IdeServices.

slluis committed:
Track API changes

All services are now referenced from the IdeServices class

Therzok commented on main/src/addins/MacPlatform/Dialogs/MacOpenFileDialogHandler.cs:
```diff
@@ -220,7 +220,7 @@ static bool FillViewers (List<FileViewer> currentViewers, NSPopUpButton button,
 			}
 			
-			foreach (var vw in DisplayBindingService.GetFileViewers (filename, null)) {
+			foreach (var vw in IdeServices.DisplayBindingService.GetFileViewers (filename, null).Result) {
```
Can we make this async somehow?

slluis replied:
I guess it can be done, but it requires changes in the dialog that are better done in another PR.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/BasicServiceProvider.cs:
```diff
@@ -0,0 +1,214 @@
+			CheckValid ();
+			Task<object> currentInitTask = null;
+
+			lock (services) {
```
Does this mean we can't initialize services in parallel?

Therzok replied:
Ah, nevermind, we can.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/BasicServiceProvider.cs:
```diff
@@ -0,0 +1,214 @@
+						services.Add (service);
+						servicesByType [typeof (T)] = service;
+						if (service is IService serviceInstance) {
+							var timer = Stopwatch.StartNew ();
```
Can we not use InstrumentationService here?

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/BasicServiceProvider.cs:
```diff
@@ -0,0 +1,214 @@
+							var completionTask = new TaskCompletionSource<object> ();
+							initializationTasks [service] = completionTask;
+							serviceInstance.Initialize (this).ContinueWith (t => {
+								LoggingService.LogInfo ($"Service {typeof (T)} initialized in {timer.ElapsedMilliseconds} ms.");
```
Timer can be skewed based on threadpool availability?

slluis replied:
Could be, but threadpool availability is just one more factor that can slow down service initialization, I don't think we need to special case it.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/BasicServiceProvider.cs:
```diff
@@ -0,0 +1,214 @@
+	/// </summary>
+	public class BasicServiceProvider : ServiceProvider
+	{
+		List<object> services = new List<object> ();
```
```suggestion
		readonly List<object> services = new List<object> ();
```

To ensure lock integrity.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/BasicServiceProvider.cs:
```diff
@@ -0,0 +1,214 @@
+				disposing = true;
+			}
+
+			await Task.WhenAll (list.OfType<IService> ().Select (s => s.Dispose ()));
```
```suggestion
			await Task.WhenAll (list.OfType<IService> ().Select (s => s.Dispose ()).ToArray ());
```

Therzok replied:
Actually, this is Dispose. Not that necessary.

Therzok committed:
Update main/src/core/MonoDevelop.Core/MonoDevelop.Core/BasicServiceProvider.cs

Co-Authored-By: slluis <llsan@microsoft.com>

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/BasicServiceProvider.cs:
```diff
@@ -0,0 +1,214 @@
+		/// <param name="action">Action to run</param>
+		/// <typeparam name="T">Service type</typeparam>
+		/// <remarks>This method does not cause the initialization of the service.</remarks>
+		public override void WhenServiceInitialized<T> (Action<T> action)
```
As stated on the other PR, I think we need to pass an IDisposable registration which would remove the callback when disposed.

That way we can't run a document's registration after the document is disposed.

slluis replied:
Fixed by https://github.com/mono/monodevelop/pull/7172/commits/36b90fc92f1649013378b160ec602aab0111250f.

slluis committed:
Add support for unregistering service initialization callbacks

Also added unit tests for the BasicServiceProvider class.

rodrmoya commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/Ide.cs:
```diff
@@ -203,22 +196,30 @@ static IdeApp ()
+			await Runtime.GetService<TextEditorService> ();
+			await Runtime.GetService<NavigationHistoryService> ();
+			await Runtime.GetService<DisplayBindingService> ();
+
```
Can these services be initialised in parallel? If so, maybe this could be changed to ```await Task.WhenAll (Runtime.GetService<TextEditorService> (), await Runtime.GetService<NavigationHistoryService> (), ...)```?

slluis replied:
Done.

rodrmoya commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/Ide.cs:
```diff
@@ -203,22 +196,30 @@ static IdeApp ()
+			await Runtime.GetService<DesktopService> ();
+			await Runtime.GetService<FontService> ();
+			await Runtime.GetService<TaskService> ();
+
```
ditto here, if they don't have dependencies between each other?

Therzok replied:
They're UI thread bound, so I think not. We're using the UI thread scheduler for this.

slluis replied:
I'll add a WhenAll, although there is little chance of parallelization since they run on the UI thread. In any case, a WhenAll won't hurt.

slluis replied:
Fixed.

rodrmoya commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/Ide.cs:
```diff
@@ -334,6 +344,14 @@ static bool FMOpenTimerExpired (Action timeCompletion)
+		{
+			Initialized += (sender, e) => {
+				if (!Ide.WelcomePage.WelcomePageService.HasWindowImplementation)
+					Workbench.Present ();
```
Not sure what this is supposed to do, but if you want to check if the Welcome Window is visible, there is a `WelcomeWindowVisible` property in `WelcomePageService`

Therzok replied:
If we don't have a welcome window, show the workbench.

rodrmoya replied:
Right 😂but when is this called?

Therzok replied:
Oh, I meant this might be to circumvent the usage of the welcome page.

slluis committed:
Service improvements

Converted HelpService and HelpOperations to the new service model.
Initialize IDE services in parallel.

slluis committed:
Merge remote-tracking branch 'origin/master' into new-service-model

Therzok commented:
@slluis one idea comes to mind with this PR.

Can we remove startup handlers and add on service init handlers?

So addins can plug in handling when a service is initialized, not on IDE startup.

slluis commented:
> @slluis one idea comes to mind with this PR.
> 
> Can we remove startup handlers and add on service init handlers?
> 
> So addins can plug in handling when a service is initialized, not on IDE startup.

Maybe, but in practice it wouldn't make much of a difference, since all services end being initialized at startup.

slluis committed:
Merge remote-tracking branch 'origin/master' into new-service-model

slluis committed:
Merge remote-tracking branch 'origin/master' into new-service-model

slluis committed:
Merge remote-tracking branch 'origin/master' into new-service-model

Therzok commented:
Was already merged to master-vnext.
